### PR TITLE
[RFC] vim-patch:7.4.579

### PR DESCRIPTION
**Pull request by glts**
```
Problem:    Wrong cursor positioning when 'linebreak' is set and lines wrap.
Solution:   (Christian Brabandt)
```

Repro steps at https://groups.google.com/d/msg/vim_dev/Eh3N9L68Ajw/4dB5x1RTQJQJ

https://github.com/vim/vim/commit/v7-4-579

Original patch:

``` diff
diff --git a/src/nvim/charset.c b/src/nvim/charset.c
--- a/src/nvim/charset.c
+++ b/src/nvim/charset.c
@@ -1178,28 +1178,50 @@ win_lbr_chartabsize(wp, line, s, col, he
     added = 0;
     if ((*p_sbr != NUL || wp->w_p_bri) && wp->w_p_wrap && col != 0)
     {
-   numberextra = win_col_off(wp);
+   colnr_T sbrlen = 0;
+   int numberwidth = win_col_off(wp);
+
+   numberextra = numberwidth;
    col += numberextra + mb_added;
    if (col >= (colnr_T)W_WIDTH(wp))
    {
        col -= W_WIDTH(wp);
        numberextra = W_WIDTH(wp) - (numberextra - win_col_off2(wp));
-       if (numberextra > 0)
+       if (col >= numberextra && numberextra > 0)
        col %= numberextra;
        if (*p_sbr != NUL)
        {
-       colnr_T sbrlen = (colnr_T)MB_CHARLEN(p_sbr);
+       sbrlen = (colnr_T)MB_CHARLEN(p_sbr);
        if (col >= sbrlen)
            col -= sbrlen;
        }
-       if (numberextra > 0)
+       if (col >= numberextra && numberextra > 0)
        col = col % numberextra;
+       else if (col > 0 && numberextra > 0)
+       col += numberwidth - win_col_off2(wp);
+
+       numberwidth -= win_col_off2(wp);
    }
-   if (col == 0 || col + size > (colnr_T)W_WIDTH(wp))
+   if (col == 0 || col + size + sbrlen > (colnr_T)W_WIDTH(wp))
    {
        added = 0;
        if (*p_sbr != NUL)
-       added += vim_strsize(p_sbr);
+       {
+       if (size + sbrlen + numberwidth > (colnr_T)W_WIDTH(wp))
+       {
+           /* calculate effective window width */
+           int width = (colnr_T)W_WIDTH(wp) - sbrlen - numberwidth;
+           int prev_width = col ? ((colnr_T)W_WIDTH(wp) - (sbrlen + col)) : 0;
+           if (width == 0)
+           width = (colnr_T)W_WIDTH(wp);
+           added += ((size - prev_width) / width) * vim_strsize(p_sbr);
+           if ((size - prev_width) % width)
+           /* wrapped, add another length of 'sbr' */
+           added += vim_strsize(p_sbr);
+       }
+       else
+           added += vim_strsize(p_sbr);
+       }
        if (wp->w_p_bri)
        added += get_breakindent_win(wp, line);

diff --git a/src/nvim/screen.c b/src/nvim/screen.c
--- a/src/nvim/screen.c
+++ b/src/nvim/screen.c
@@ -2842,6 +2842,9 @@ win_line(wp, lnum, startrow, endrow, noc
     unsigned   off;            /* offset in ScreenLines/ScreenAttrs */
     int        c = 0;          /* init for GCC */
     long   vcol = 0;       /* virtual column (for tabs) */
+#ifdef FEAT_LINEBREAK
+    long   vcol_sbr = -1;      /* virtual column after showbreak */
+#endif
     long   vcol_prev = -1;     /* "vcol" of previous character */
     char_u *line;          /* current line */
     char_u *ptr;           /* current position in "line" */
@@ -3759,6 +3762,7 @@ win_line(wp, lnum, startrow, endrow, noc
            n_extra = (int)STRLEN(p_sbr);
            char_attr = hl_attr(HLF_AT);
            need_showbreak = FALSE;
+           vcol_sbr = vcol + MB_CHARLEN(p_sbr);
            /* Correct end of highlighted area for 'showbreak',
             * required when 'linebreak' is also set. */
            if (tocol == vcol)
@@ -4516,9 +4520,17 @@ win_line(wp, lnum, startrow, endrow, noc
        if (c == TAB && (!wp->w_p_list || lcs_tab1))
        {
            int tab_len = 0;
+           long vcol_adjusted = vcol; /* removed showbreak length */
+#ifdef FEAT_LINEBREAK
+           /* only adjust the tab_len, when at the first column
+            * after the showbreak value was drawn */
+           if (*p_sbr != NUL && vcol == vcol_sbr && wp->w_p_wrap)
+           vcol_adjusted = vcol - MB_CHARLEN(p_sbr);
+#endif
            /* tab amount depends on current column */
            tab_len = (int)wp->w_buffer->b_p_ts
-                   - vcol % (int)wp->w_buffer->b_p_ts - 1;
+                   - vcol_adjusted % (int)wp->w_buffer->b_p_ts - 1;
+
 #ifdef FEAT_LINEBREAK
            if (!wp->w_p_lbr || !wp->w_p_list)
 #endif
diff --git a/src/nvim/version.c b/src/nvim/version.c
--- a/src/nvim/version.c
+++ b/src/nvim/version.c
@@ -742,6 +742,8 @@ static char *(features[]) =
 static int included_patches[] =
 {   /* Add new patch number below this line */
 /**/
+    579,
+/**/
     578,
 /**/
     577,
```

### CODE REVIEW

File: src/nvim/charset.c [hunk 1]
```diff
@@ -1049,29 +1049,54 @@ int win_lbr_chartabsize(win_T *wp, char_u *line, char_u *s, colnr_T col, int *he
...
-        if (col >= sbrlen)
+        sbrlen = (colnr_T)MB_CHARLEN(p_sbr);
+        if (col >= sbrlen) {
           col -= sbrlen;
+        }
       }
-      if (numberextra > 0) {
-        col = col % numberextra;
+      if (col >= numberextra && numberextra > 0) {
+        col %= numberextra;
```
Comment (fwalch) [inline src/nvim/charset.c]: Huh, never seen that before :-D

File: src/nvim/charset.c [hunk 2]
```diff
@@ -1049,29 +1049,54 @@ int win_lbr_chartabsize(win_T *wp, char_u *line, char_u *s, colnr_T col, int *he
...
+      } else if (col > 0 && numberextra > 0) {
+        col += numberwidth - win_col_off2(wp);
       }
+
+      numberwidth -= win_col_off2(wp);
     }
 
-    if ((col == 0) || (col + size > (colnr_T)wp->w_width)) {
+    if (col == 0 || (col + size + sbrlen > (colnr_T)wp->w_width)) {
```
Comment (fwalch) [inline src/nvim/charset.c]: Apparently we introduced braces around the conditions somewhere? Not sure if we have a style rule somewhere, but the way you changed it makes sense to me.

**Comment by fwalch** [comment]
LGTM.

File: src/nvim/charset.c (hunk 2 above)
Comment (glts) [inline src/nvim/charset.c]: I don't remember seeing anything about this in the style guide actually. I applied the rule of thumb 'is the meaning clear at first glance without brackets?', and my answer was yes for the expression on the left.

File: src/nvim/charset.c (hunk 1 above)
Comment (glts) [inline src/nvim/charset.c]: The same statement `col %= numberextra;` is also used a few lines before that, which is why I changed it.

**Comment by glts** [comment]
Thanks for reviewing!

**Comment by justinmk** [comment]
Merged, thanks!

**Pull request closed without merging**